bitbots_ball_filter:
  ros:
    camera_info_subscribe_topic:
      type: string
      default_value: 'camera/camera_info'
      description: 'Topic of the camera info'
      read_only: True

    ball_subscribe_topic:
      type: string
      default_value: 'balls_relative'
      read_only: true
      description: 'Topic to subscribe to for ball detections'

    ball_position_publish_topic:
      type: string
      default_value: 'ball_position_relative_filtered'
      read_only: true
      description: 'Topic to publish the filtered ball position'
    
    ball_filter_reset_service_name: 
      type: string
      default_value: 'ball_filter_reset'
      read_only: true
      description: 'Service to reset the ball filter'

  filter:
    frame:
      type: string
      default_value: 'map'
      description: 'Frame to filter the ball in'
      read_only: true
      validation:
        one_of<>: [['map', 'odom']]

    rate:
      type: int
      default_value: 62
      description: 'Filtering rate in Hz'
      read_only: true
      validation: 
        bounds<>: [0, 100]  

    covariance:
      process_noise: 
        type: double
        default_value: 0.005
        description: 'Noise which is added to the estimated position without new measurements'
        validation:
          bounds<>: [0.0, 1.0]

      measurement_uncertainty:
        type: double
        default_value: 0.5
        description: 'Ball measurement certainty in filter'
        validation:
          bounds<>: [0.0, 1.0]
      
      distance_factor: 
        type: double
        default_value: 0.02
        description: 'Factor to increase the noise if the ball measurement is further away. This also depends on the reference distance.'
        validation:
          bounds<>: [0.0, 1.0]
  
      negative_observation:
        value:
          type: double
          default_value: 1.5
          description: 'Value by which the covariance is increased if we have a negative observation, meaning we did not see the ball if we should have seen it.'
          validation:
            bounds<>: [0.0, 10.0]

        ignore_border:
          type: double
          default_value: 0.1
          description: 'Border as a fraction of the image height and width. We do not consider the negative observation if they are close to the border of the image, as we might not see the ball there, when it is partially occluded.'
          validation:
            bounds<>: [0.0, 1.0]

    tracking:
      ignore_measurement_threshold:
        type: double
        default_value: 6.0
        description: 'We calculate a distance around our filter state over which we ignore measurements. This threshold is based on the covariance of the filter scaled by this factor. This is used to ignore false positives if we already have a good estimation.'
        validation:
          bounds<>: [0.0, 10.0]
